File: src/main/java/org/c3lang/intellij/stubs/StubStreamExtensions.java
package org.c3lang.intellij.stubs;

import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import com.intellij.util.io.DataInputOutputUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public final class StubStreamExtensions
{
	private StubStreamExtensions() {}

	public static void writeNullableUTFFast(@NotNull StubOutputStream s, @Nullable String arg) throws IOException
	{
		DataInputOutputUtil.writeNullable(s, arg, s::writeUTFFast);
	}

	public static @Nullable String readNullableUTFFast(@NotNull StubInputStream s) throws IOException
	{
		return DataInputOutputUtil.readNullable(s, s::readUTFFast);
	}

	public static void writeList(@NotNull StubOutputStream s, @NotNull List<String> list) throws IOException
	{
		s.writeVarInt(list.size());
		for (String item : list)
		{
			s.writeUTFFast(item);
		}
	}

	public static @NotNull List<String> readList(@NotNull StubInputStream s) throws IOException
	{
		int size = s.readVarInt();
		List<String> result = new ArrayList<>(size);
		for (int i = 0; i < size; i++)
		{
			result.add(s.readUTFFast());
		}
		return result;
	}
}

File: src/main/java/org/c3lang/intellij/stubs/StubStreamExtensions.kt
package org.c3lang.intellij.stubs

import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import com.intellij.util.io.DataInputOutputUtil

fun StubOutputStream.writeNullableUTFFast(arg: String?) {
    DataInputOutputUtil.writeNullable(this, arg, this::writeUTFFast)
}

fun StubInputStream.readNullableUTFFast(): String? {
    return DataInputOutputUtil.readNullable(this, this::readUTFFast)
}

fun StubOutputStream.writeList(list: List<String>) {
    writeVarInt(list.size)

    for (string in list) {
        writeUTFFast(string)
    }
}

fun StubInputStream.readList(): List<String> {
    val size = readVarInt()

    return (0 until size).map {
        readUTFFast()
    }.toList()
}
